samples/java-protobuf-tracing/src/main/java/com/example/ControllerAction.java

```diff
@@ -0,0 +1,137 @@
+    HttpRequest request = HttpRequest.newBuilder()
+            .uri(URI.create(url))
+            .build();
+    return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+  }
+
+  private String callSyncService(Tracer tracer) {
+    Span span = tracer
+            .spanBuilder("loreipsumendpoint")
+            .setParent(actionContext().metadata().traceContext().asOpenTelemetryContext())
+            .startSpan();
+    span.setAttribute("attribute1", "value1");
+
+    String result = "";
+    //scope must be closed. try-with-resources will automatically close it.
+    try (Scope scope = span.makeCurrent()) {
+      //Sync call to external service
+      HttpResponse<String> response = callSyncService();
+      result = response.body();
+      span.setAttribute("result", result);
+    } catch (IOException | InterruptedException e) {
+      result = e.getMessage();
+      span.setStatus(StatusCode.ERROR, result);
+    } finally {
+      span.end();
+    }
+    return result;
+  }
+
+
+    @Override
+  public Effect<ControllerActionApi.MessageResponse> callAsyncEndpoint(Empty empty) {
+      // tag::get-tracer[]
+      Optional<Tracer> tracerOpt = actionContext().getOpenTelemetryTracer();
+      // end::get-tracer[]
+      CompletableFuture<HttpResponse<String>> futureResponse;
+
+    if(tracerOpt.isPresent()) {
+      futureResponse = callAsyncService(tracerOpt.get());
+    } else {
+      futureResponse = callAsyncService();
+    }
+    CompletionStage<ControllerActionApi.MessageResponse> asyncComputation = futureResponse.toCompletableFuture().thenApply(response -> {
+      return ControllerActionApi.MessageResponse.newBuilder().setMessage(response.body()).build();
+    });
+    return effects().asyncReply(asyncComputation);
+  }
+
+
+  private CompletableFuture<HttpResponse<String>> callAsyncService(){
+    HttpRequest httpRequest = HttpRequest.newBuilder()
+            .uri(URI.create(url))
+            .build();
+    //Async call to external service
+    return httpClient.sendAsync(httpRequest,
+            HttpResponse.BodyHandlers.ofString());
+
+  }
+
+
+  //  tag::create-close-span[]
+  private CompletableFuture<HttpResponse<String>> callAsyncService(Tracer tracer) {
+  // end::create-close-span[]
+
+    // tag::create-close-span[]
+    Span span  = tracer 
+            .spanBuilder("loreipsumendpoint")
+            .setParent(actionContext().metadata().traceContext().asOpenTelemetryContext())// <1>
+            .startSpan(); // <2>
+    span.setAttribute("attribute1", "value1");// <3>
+
+    CompletableFuture<HttpResponse<String>> responseFuture = callAsyncService();
+
+    try (Scope scope = span.makeCurrent()) {// <4>
+      responseFuture.thenAccept(response -> {
+        String responseBody = response.body();
+        span.setAttribute("result", responseBody);// <5>
+      }).exceptionally(ex -> {
+        span.setStatus(StatusCode.ERROR, ex.getMessage());
+        return null;
```

Comment:
This will be a problem. 

In case of an error, it will later cause a NPE because we are recovering the failed CompletionState with a successful one but if a null object. 

Instead, we should preserve the failure and later, in `public Effect<ControllerActionApi.MessageResponse> callAsyncEndpoint(Empty empty)`, we need to either return a `effects().reply` or a `effects().error`

Reply:
Good catch. 
One thing. I've been looking at this and the only thing we can do is use `.exceptionally` and inside create a `MessageResponse` with the error or throw our own exception. I understand we can not use `effects().error` once we enter the monad-like `CompletableFuture`. That's why we use the `asyncReply` in the first place. I thought that we should create an `asyncError` but then I started to think that it will be just like `asyncReply`. WDYT?

Reply:
So, just for the record. We got a call and things are now much clear.

I was wrong about the NPE. The `null` is never returned because this method is always returning the original `responseFuture`.

The code run inside the try, are just call backs for the successful and failure case.

Reply:
Guys, why, I have this feeling that it will not work as expected when we hit it with many concurrent requests, I'm worried about this part:
`try (Scope scope = span.makeCurrent()) {// <4>`

My implementation is a bit different: 
```
Optional<Span> span = startSpan("slack-api-post-new-cfp");
        return callAsyncService().thenApply(response -> {
            span.ifPresent(Span::end);
            return response;
          }
        );
```

```
private Optional<Span> startSpan(String spanName) {
    var otelCurrentContext = actionContext().metadata().traceContext().asOpenTelemetryContext();
    Optional<Span> span = actionContext().getOpenTelemetryTracer().map(tracer -> tracer
      .spanBuilder(spanName)
      .setParent(otelCurrentContext)
      .setSpanKind(SpanKind.CLIENT)
      .startSpan());
    return span;
  }
```

Could also include the `exceptionally` part, but here I don't have to worry about any current otel context.

Reply:
I added a comment to a resolved conversation, unresolving it.

Reply:
I see the changes now; it's better, but still, I would remove the `try` statement and simplify this code.

Reply:
Let me give it a try. No pun intended.

Reply:
I'm gladly surprised. I did try this a couple of times without the `makeCurrent()` and couldn't make it work. Looked at the docs and didn't understand why. In any case it just works. So yeah, much better without it. Not sure it would have an impact because it was just writing but nevertheless, in the best possible scenario it was noise. 

Removed now.